### ISSUE TITLE
Frequent "Invalid signature for user or client"

### ISSUE THREAD

**Issue by phene** [sethvargo/chef-api#32]
I cannot reproduce it consistently, but I frequently get transient errors for the api key when connecting to the chef server.  I've verified that I'm using the correct key, and there are no time sync issues between the servers.

**Comment by sethvargo** [comment]
@phene this usually means times are off. Are you running your own Chef server or ...? What versions? There is not nearly enough information here to help me adequately diagnose the problem.

**Comment by phene** [comment]
As stated before, I verified the times were in sync as that was my first suspicion.  I am running my own chef server on version 11.12.8.  I agree that there's not a lot of info to go on here, but I figured it would be worth posting this in case others ran into the same issue, which may be indicative of a problem in the library.

**Comment by sethvargo** [comment]
@phene without a solid reproduction case and no other user's reporting this problem, the only conclusion is something amiss in your Chef server installation. If you are able to provide more information in the future, we can re-open this issue.

**Comment by phene** [comment]
Here's a related ticket https://github.com/chef/chef-server/issues/5

**Comment by sethvargo** [comment]
@phene the fact that it's intermittent makes me feel as though it's definitely a server error and not a client error. I'm not denying this is a bug in Chef API, but I can't reproduce the issue, and you are unable to provide a solid reproduction case. Without more information, there's nothing I can do. 

I use the Chef API gem very heavily and have never experienced this issue with Hosted Chef, which also leads me to believe there is an issue with your local Chef Server installation. The [authentication module](https://github.com/sethvargo/chef-api/blob/master/lib/chef-api/authentication.rb) is pretty straight forward and follows the exact algorithm for generating the signed header auth as presented in the Chef documentation and Chef itself. ChefAPI has a very verbose debug mode that will output all the signed headers, etc, so you could compare those to what Chef is generating.

I'm not trying to be dismissive, but I do not really know what you want me to do here... I can't fix an issue that even you yourself are unable to reproduce reliably :smile:.

**Comment by skierkowski** [comment]
FYI. Last night I was seeing the same issue. My tests were getting ~15% failure rate and today ~1% randomly across all API endpoints. Hosted Chef support verified that this was a server-side issue, and they will be diving deeper on Monday.

Having said that, would make sense to have retry logic in chef-api?

**Comment by sethvargo** [comment]
@skierkowski I'm usually :-1: on retries in a library because the end-user likely wants more control over the functionality. So, instead of:

```ruby
Thing.all(retry: 3, retry_on: [ResourceNotFound], timeout: 50)
```

and supporting a complex API with every option a user could ever want, I prefer the API to just be:

```ruby
Thing.all
```

and the implementing library can wrap that in whatever retry logic it wants:

```ruby
def retry(&block)
  Timeout.timeout(50) do
    yield
  end
rescue MyCustomException
  do_something
rescue MyOtherCustomException
  do_something_else
end

retry do
  Thing.all
end
```

**Comment by skierkowski** [comment]
Thanks. That makes sense. I'll likely wait it out for the server-side pieces to get fixed before I dive into (dreaded) retries.

**Comment by wdeviers** [comment]
I know this has been closed, but I see the same thing. I assumed it was a problem server-side as well, I've been working with Chef Enterprise support for a few weeks on this issue and am 100% unable to reproduce the problem on our Enterprise server install. I wrote a threaded wrapper around a shell/curl script that runs similar API calls. I can do 10k API calls over an hour and get no signature problems, but as soon as I use something with chef-api, the signature problems start showing up in 20-50% of cases, roughly.

I've included chef-api as part of a hack-and-slash Sinatra project.  I'm unsure yet where in the stack the problem falls, but I *think* I definitively say that it's not chefd.

**Comment by spuder** [comment]
I can reproduce this rather consistently and I have captured tcp dumps of both successful and unsuccessful api calls. While I do suspect this is a bug in chef server, because it only affects api calls with this ruby gem, I'm posting my data here instead of https://github.com/chef/chef-server/issues/5 since 

**Symptoms:**

- Fails about 5 to 10 times out of 50 attempts. 
- When the ruby script fails, running knife commands in a separate terminal will also fail for a few seconds. (I've never had a knife command fail otherwise)
- Fails more often on longer "fetches" (like getting cookbook versions)

I don't consider myself a developer, but here is some ruby which reproduces the problem quite nicely. 

sausage.rb
```ruby
require 'chef-api'
include ChefAPI::Resource

sausage = ChefAPI::Connection.new(
  endpoint: 'https://api.opscode.com/organizations/spuder',
  client:   'spuder',
  key:      '~/.chef/sausage.pem'
)

@aCookbook   = 'elasticsearch'
@theEnvironments = sausage.environments.list
@aDatabagItem = 'foo'

@cookbooks = Hash.new({ value: 0 })
@cookbookAttribHash = Hash.new({ value: 0 })

def cookbookVersions(organization)
  @theEnvironments.each do |anEnvironment|
    cookbook_version = organization.environments.fetch(anEnvironment).cookbook_versions[@aCookbook]
    @cookbooks[anEnvironment] = { label: anEnvironment, value: cookbook_version }
  end
  return @cookbooks.values
end

def environmentAttributes(organization)
  @theEnvironments.each do |anEnvironment|
    if organization.environments.fetch(anEnvironment).default_attributes[@aCookbook]
      cookbook_version = organization.environments.fetch(anEnvironment).default_attributes[@aCookbook]['foo']
      @cookbookAttribHash[anEnvironment] = { label: anEnvironment, value: cookbook_version }
    else
      @cookbookAttribHash[anEnvironment] = { label: anEnvironment, value: nil }
    end
  end
  return @cookbookAttribHash.values
end


puts cookbookVersions(sausage)
puts environmentAttributes(sausage)

```

You can replace the organization, key and client information, then test like so: 

    F=0; for i in {1..50}; do echo "Executing run $i"; ruby sausage.rb; if [ $? !=0 ];then F=$((F+1));fi; done; echo "Failed $F times"

In my last test, I had 14 failures out of 50 attempts. 

Here is what good output should look like. 
The first 9 rows show the version of 'elasticsearch' in each of my environments
The second 9 rows show the value of a default attribute in an environment. 

```bash
foo@bar-MBP:$ ruby sausage.rb
{:label=>"_default", :value=>nil}
{:label=>"datacenter01", :value=>nil}
{:label=>"datacenter02", :value=>"<= 1.1.1"}
{:label=>"datacenter03", :value=>"~> 2.2.2"}
{:label=>"datacenter04", :value=>"~> 9.9.9"}
{:label=>"datacenter05", :value=>nil}
{:label=>"datacenter06", :value=>nil}
{:label=>"datacenter07", :value=>nil}
{:label=>"datacenter08", :value=>nil}
{:label=>"_default", :value=>nil}
{:label=>"datacenter01", :value=>nil}
{:label=>"datacenter02", :value=>nil}
{:label=>"datacenter03", :value=>nil}
{:label=>"datacenter04", :value=>nil}
{:label=>"datacenter05", :value=>nil}
{:label=>"datacenter06", :value=>20160808.2}
{:label=>"datacenter07", :value=>nil}
{:label=>"datacenter08", :value=>nil}
```

A failure will output the following

```bash

foo@bar-MBP:$ ruby sausage.rb
{:label=>"_default", :value=>nil}
{:label=>"datacenter01", :value=>nil}
{:label=>"datacenter02", :value=>"<= 1.1.1"}
{:label=>"datacenter03", :value=>"~> 2.2.2"}
{:label=>"datacenter04", :value=>"~> 9.9.9"}
{:label=>"datacenter05", :value=>nil}
{:label=>"datacenter06", :value=>nil}
{:label=>"datacenter07", :value=>nil}
{:label=>"datacenter08", :value=>nil}
/Users/foo/.rvm/gems/ruby-2.2.0/gems/chef-api-0.5.0/lib/chef-api/connection.rb:413:in `error': The Chef Server requires authorization. Please ensure you have specified the correct client name and private key. If this error continues, please verify the given client has the proper permissions on the Chef Server. (ChefAPI::Error::HTTPUnauthorizedRequest)

    {"error":["Invalid signature for user or client 'spuder'"]}
	from /Users/foo/.rvm/gems/ruby-2.2.0/gems/chef-api-0.5.0/lib/chef-api/connection.rb:282:in `block in request'
	from /Users/foo/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/net/http.rb:853:in `start'
	from /Users/foo/.rvm/gems/ruby-2.2.0/gems/chef-api-0.5.0/lib/chef-api/connection.rb:268:in `request'
	from /Users/foo/.rvm/gems/ruby-2.2.0/gems/chef-api-0.5.0/lib/chef-api/connection.rb:106:in `get'
	from /Users/foo/.rvm/gems/ruby-2.2.0/gems/chef-api-0.5.0/lib/chef-api/resources/base.rb:252:in `fetch'
	from sausage.rb:30:in `block in environmentAttributes'
	from sausage.rb:28:in `each'
	from sausage.rb:28:in `environmentAttributes'
	from sausage.rb:42:in `<main>'
```

I have captured wireshark dumps and filtered to just a couple hundred packets that show success and failure scenarios. Unfortunately, chef appears to be preferring ECDHE, which from my research uses perfect forward secrecy, and is impossible to decrypt, even with my chef private and public keys. (If anyone knows a way to get decrypted tcp dumps with chef, please let me know). 

    Cipher Suite: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (0xc02f)

Here are the wireshark dumps. Two are on my home network, and two are while tethered to my cellphone 4g connection. 

https://dl.dropboxusercontent.com/u/460997/chef-bug/good-sausage-filtered.pcapng
https://dl.dropboxusercontent.com/u/460997/chef-bug/good-sausage-4g-filtered.pcapng
https://dl.dropboxusercontent.com/u/460997/chef-bug/bad-sausage-filtered.pcapng
https://dl.dropboxusercontent.com/u/460997/chef-bug/bad-sausage-4g-filtered.pcapng

Interesting data points
- Every tcp dump has packets that arrived out of order, and also dropped packets
- There are a lot of "Server Hello" transactions which seems to indicate the gem is opening a new connection for every api call. (I haven't worked much with API's and development to know if that is normal.)



**Environment**
Mac Yosemite
Chef Development Kit Version: 0.3.5
Hosted chef (with 0 nodes, 1 user,  ~ 8 environments, 2 organizations)

**Comment by spuder** [comment]
I tested this again against a local chef server (instead of hosted), and I can reproduce it here too. (Though at a much lower frequency). 

- Different user
- Different key
- no ssl 

```ruby
chef_lab = ChefAPI::Connection.new(
  endpoint: 'https://devops-chef.example.com/organizations/dev',
  client:   'foo',
  key:      '~/.chef/foo.pem',
  ssl_verify: false
)
```

chef-server 12.0.8
chef-client: 12.3.0

**Comment by stevendanna** [comment]
@spuder Thanks for the detailed investigation!

@sethvargo As a heads up, I'll take a look at this from the chef-server side this week.

**Comment by sethvargo** [comment]
Hi @stevendanna 

Thanks!

**Comment by stevendanna** [comment]
OK.  I don't think this is a server bug.  In the cases I was able to reproduce, the following was the problem: https://github.com/sethvargo/chef-api/pull/39

It's interesting to me that some reporters where seeing this frequently, since it should occur pretty randomly if that is the only cause of it.  My guess is that the reporters here were experiencing some combination of the bug I posted above and intermittent server load issues.
